File: p051/e51.c
#include "stdio.h"
#include "stdlib.h"

#include <math.h>

/*
 * #51
 * By replacing the 1st digit of the 2-digit number *3, it turns
 * out that six of the nine possible values: 13, 23, 43, 53, 73,
 * and 83, are all prime.
 *
 * By replacing the 3rd and 4th digits of 56**3 with the same
 * digit, this 5-digit number is the first example having seven
 * primes among the ten generated numbers, yielding the family:
 * 56003, 56113, 56333, 56443, 56663, 56773, and 56993.
 * Consequently 56003, being the first member of this family, is
 * the smallest prime with this property.
 *
 * Find the smallest prime which, by replacing part of the number
 * (not necessarily adjacent digits) with the same digit, is part
 * of an eight prime value family.
 */

int prime(int n)
{
	int q;
	for (q = 2; q <= floor(sqrt(n)); q++)
		if (n % q == 0)
			return 0;
	return 1;
}

int nextprime(int n)
{
	for (n++; ! prime(n); n++);
	return n;
}

int replace(int mask, int n, int r)
{
	int ld;
	if (n == 0) return 0;
	ld = mask & 1 ? r : n % 10;
	return replace(mask >> 1, n / 10, r) * 10 + ld;
}

int digits(int n)
{
	int digits;
	for (digits = 0; n > 0; n /= 10) digits++;
	return digits;
}

int countto(int n)
{
	int ds, pow2;
	ds = digits(n);
	for (pow2 = 1; ds > 0; ds--) pow2 *= 2;
	return pow2;
}

void printperms(int n, int r)
{
	int m;
	for (m = 0; m < countto(n); m++)
		printf("%04d\n", replace(m, n, r));
}

int primefamily(int m, int n)
{
	int r, ps, replaced;
	ps = 0;
	for (r = 0; r < 10; r++) {
		replaced = replace(m, n, r);
		if (digits(replaced) != digits(n))
			continue;
		if (prime(replaced))
			ps++;
	}
	return ps;
}

int main(int argc, char **argv)
{
	int p, m, f, t;

	if (argc != 2) {
		fprintf(stderr, "usage: e51 familysize\n");
		return EXIT_FAILURE;
	}
	t = atoi(argv[1]);

	for (p = 13;; p = nextprime(p))
		for (m = 1; m < countto(p); m++) {
			f = primefamily(m, p);
			if (f >= t) {
				printf("%d\t0x%x\t%d\t%d\n", p, m, replace(m, p, 0), replace(m, p, 1));
				return EXIT_SUCCESS;
			}
		}
	return EXIT_SUCCESS;
}

File: p051/e51.fs
\ #51
\ By replacing the 1st digit of the 2-digit number *3, it turns
\ out that six of the nine possible values: 13, 23, 43, 53, 73,
\ and 83, are all prime.
\
\ By replacing the 3rd and 4th digits of 56**3 with the same
\ digit, this 5-digit number is the first example having seven
\ primes among the ten generated numbers, yielding the family:
\ 56003, 56113, 56333, 56443, 56663, 56773, and 56993.
\ Consequently 56003, being the first member of this family, is
\ the smallest prime with this property.
\
\ Find the smallest prime which, by replacing part of the number
\ (not necessarily adjacent digits) with the same digit, is part
\ of an eight prime value family.

\ Prime numbers
: isqrt ( d -- root ) s>f fsqrt f>s ;
: divides? ( s q -- ? ) mod 0= ;
: prime? ( n -- ? )
    dup isqrt 1+ 2 ?do
        dup i divides? if unloop drop false exit then
    loop drop true ;
: nextp ( n -- next ) begin 1+ dup prime? until ;

\ Replacing digits based on a mask
variable mask
variable replacement
variable acc
variable mult
: lastdigit ( n -- l ) mask @ 1 and if drop replacement @ else 10 mod then ;
: shiftmask ( -- ) mask @ 2/ mask ! ;
: +mult ( -- ) mult @ 10 * mult ! ;
: nextn ( n -- n' ) 10 / ;
: next ( n -- n ) shiftmask +mult nextn ;
: accdigit ( n -- ) lastdigit mult @ * acc +! ;
: replaced ( n -- n' )
	0 acc ! 1 mult !
	begin dup accdigit next dup 0= until
	drop acc @ ;

\ Counting the primes in the family
: rmask! ( r m -- ) mask ! replacement ! ;
variable primes
: digits ( n -- ds ) 0 swap begin dup 0<> while 10 / >r 1+ r> repeat drop ;
: valid? ( p r -- ) dup prime? >r digits >r digits r> = r> and ;
: primefamily ( n m -- ps )
	0 primes ! >r
	10 0 do i j rmask! dup dup replaced valid? if 1 primes +! then loop
	rdrop drop primes @ ;

\ Checking each prime with each mask
: maskmax ( n -- m ) 1 begin over 0<> while 2* >r 10 / r> repeat nip ;
: e51 ( -- n )
	2 begin
		dup maskmax 1 do
			dup i primefamily 8 >= if 1 i rmask! replaced . unloop exit then
		loop
		nextp
	again ;
